// tera-client-interface/index.js
const net = require('net');
const EventEmitter = require('events');

class TeraClientInterfaceConnection extends EventEmitter {
    constructor(socket) {
        super();
        this.setMaxListeners(0);

        this.socket = socket;
        this.buffer = null;

        this.socket.on('data', data => {
            this.buffer = this.buffer ? Buffer.concat([this.buffer, data]) : data;

            let start = 0;
            let end = -1;
            while ((end = this.buffer.indexOf(0, start)) >= 0) {
                if (end > start) {
                    const packet = this.buffer.slice(start, end);
                    try {
                        const parsed = JSON.parse(packet);
                        this.emit('data', parsed.command, parsed.data || {});
                    } catch (e) {
                        console.log(`[proxy] Error communicating with client:`);
                        console.log(e);
                    }
                }

                start = end + 1;
            }

            this.buffer = this.buffer.slice(start);
        });

        this.socket.once('error', e => {
            this.socket = null;
            this.emit('disconnect', e);
            this.removeAllListeners();
        });

        this.socket.once('close', () => {
            this.socket = null;
            this.emit('disconnect');
            this.removeAllListeners();
        });
    }

    send(command, data = {}) {
        if (this.socket)
            this.socket.write(JSON.stringify({ command, data }) + "\x00");
    }

    destructor() {
        if (this.socket) {
            this.socket.end();
            this.socket.destroy();
            this.socket = null;
        }
    }
}

class TeraClientInterfaceServer {
    constructor(host, port, onAccept) {
        this.onAccept = onAccept;

        this.server = net.createServer(socket => this.accept(socket));
        this.server.listen(port, host);

        this.connections = new Set;
    }

    destructor() {
        this.connections.forEach(connection => connection.destructor());
        this.connections.clear();

        this.server.close();
        this.server = null;

        this.onAccept = null;
    }

    accept(socket) {
        socket.setNoDelay(true);

        const connection = new TeraClientInterfaceConnection(socket);
        this.connections.add(connection);
        socket.once('error', () => this.connections.delete(connection));
        socket.once('close', () => this.connections.delete(connection));

        this.onAccept(connection);
    }
}

module.exports = TeraClientInterfaceServer;
