// tera-client-interface/process-listener-dll-injector.js
const ProcessListener = require('./process-listener');
const { execFileSync } = require("child_process");
const path = require('path');

class ProcessListenerDLLInjector {
    constructor(Interval = 100) {
        this.listener = new ProcessListener('TERA.exe', this.handleAddedProcess, this.handleRemovedProcess, Interval);
    }

    destructor() {
        this.listener.stop();
        this.listener = null;
    }

    run() {
        this.listener.start();
    }

    setInterval(Interval) {
        this.listener.interval = Interval;
    }

    handleAddedProcess(process) {
        try {
            execFileSync(path.join(__dirname, "injector.exe"), [process.pid, path.join(__dirname, "tera-client-interface.dll"), 1]);
        } catch (e) {
            console.log(`[toolbox] ERROR: Unable to connect to game client (PID ${process.pid})!`);
            switch (e.code) {
                case 'ENOENT': {
                    console.log("[toolbox] injector.exe does not exist. It has likely been deleted by your anti-virus.");
                    console.log("[toolbox] Disable/uninstall your anti-virus or whitelist TERA Toolbox and injector.exe!");
                    break;
                }
                case 'EPERM': {
                    console.log("[toolbox] Permission to launch injector.exe denied. It has likely been blocked by your anti-virus.");
                    console.log("[toolbox] Disable/uninstall your anti-virus or whitelist TERA Toolbox and injector.exe!");
                    break;
                }
                default: {
                    switch (e.status) {
                        case 1:
                            {
                                console.log("[toolbox] Connection to game client unsuccessful.");
                                console.log("[toolbox] > Make sure that TERA Toolbox is running with Administrator privileges!");
                                console.log("[toolbox] > Disable/uninstall your anti-virus or whitelist TERA Toolbox and injector.exe!");
                                console.log("[toolbox] > Reboot your computer or kill TERA.exe processes if there are any in task manager!");
                                break;
                            }
                        default:
                            {
                                console.log("[toolbox] This is likely caused by your anti-virus interfering. Disable/uninstall it or whitelist TERA Toolbox.");
                                console.log("[toolbox] Full error message:");
                                console.log(e);
                                break;
                            }
                    }
                    break;
                }
            }
        }
    }

    handleRemovedProcess(pid) {
        // Do nothing.
    }
}

module.exports = ProcessListenerDLLInjector;
