// tera-client-interface/process-listener.js
const { snapshot } = require("process-list");

class ProcessListener {
    constructor(Name, OnAdded, OnRemoved, Interval = 500) {
        this.listedProcesses = {};
        this.name = Name.toUpperCase();
        this.onAdded = OnAdded;
        this.onRemoved = OnRemoved;
        this.interval = Interval;
        this._timeout = null;
    }

    start() {
        if (this._timeout)
            return;

        this._timeout = setTimeout(() => this.update(), this.interval);
    }

    stop() {
        if (!this._timeout)
            return;

        clearTimeout(this._timeout);
        this._timeout = null;
    }

    update() {
        snapshot("pid", "name", "path", "ppid").then(processes => {
            if (!this._timeout)
                return;

            let NewListedProcesses = {};
            for (let process of processes) {
                if (process.name.toUpperCase() === this.name && !this.listedProcesses[process.ppid]) {
                    if (!this.listedProcesses[process.pid]) {
                        this.onAdded(process);
                        NewListedProcesses[process.pid] = true;
                    } else {
                        NewListedProcesses[process.pid] = true;
                    }
                }
            }

            for (let pid in this.listedProcesses) {
                if (!NewListedProcesses[pid])
                    this.onRemoved(pid);
            }
            this.listedProcesses = NewListedProcesses;
            this._timeout = setTimeout(() => this.update(), this.interval);
        });
    }
}

module.exports = ProcessListener;
